packages/taiko-client/driver/preconf_blocks/server.go

```diff
@@ -298,6 +302,234 @@ func (s *PreconfBlockAPIServer) OnUnsafeL2Payload(
+				"hash", msg.ExecutionPayload.BlockHash.Hex(),
+				"parentHash", msg.ExecutionPayload.ParentHash.Hex(),
+				"headerHash", header.Hash().Hex(),
+				"headerParentHash", header.ParentHash.Hex(),
+			)
+		}
+	}
+
+	// Insert the preconfirmation block into the L2 EE chain.
+	if _, err := s.chainSyncer.InsertPreconfBlocksFromExecutionPayloads(
+		ctx,
+		[]*eth.ExecutionPayload{msg.ExecutionPayload},
+		false,
+	); err != nil {
+		return fmt.Errorf("failed to insert preconfirmation block from P2P network: %w", err)
+	}
+
+	// Try to import the child blocks from the cache, if any.
+	if err := s.ImportChildBlocksFromCache(ctx, msg.ExecutionPayload); err != nil {
+		return fmt.Errorf("failed to try importing child blocks from cache: %w", err)
+	}
+
+	return nil
+}
+
+// OnUnsafeL2Request implements the p2p.GossipIn interface.
+func (s *PreconfBlockAPIServer) OnUnsafeL2Request(
+	ctx context.Context,
+	from peer.ID,
+	hash common.Hash,
+) error {
+	// Ignore the message if it is from the current P2P node.
+	if from != "" && s.p2pNode.Host().ID() == from {
+		log.Debug("Ignore the message from the current P2P node", "peer", from)
+		return nil
+	}
+
+	// only respond if you are the current sequencer.
+	if err := s.checkLookaheadHandover(s.preconfOperatorAddress, s.rpc.L1Beacon.CurrentSlot()); err != nil {
+		log.Debug(
+			"Ignoring OnUnsafeL2Request, not the current sequencer",
+			"peer", from,
+			"currOperator", s.lookahead.CurrOperator.Hex(),
+			"nextOperator", s.lookahead.NextOperator.Hex(),
+			"preconfOperatorAddress", s.preconfOperatorAddress.Hex(),
+		)
+
+		return nil
+	}
+
+	log.Info("New block request from p2p network", "peer", from, "hash", hash.Hex())
+
+	block, err := s.rpc.L2.BlockByHash(ctx, hash)
+	if err != nil {
+		log.Warn("Failed to fetch block by hash", "hash", hash.Hex(), "error", err)
+		return err
+	}
+
+	var u256 uint256.Int
+	if overflow := u256.SetFromBig(block.BaseFee()); overflow {
+		log.Warn(
+			"Failed to convert base fee to uint256, skip propagating the preconfirmation block",
+			"baseFee", block.BaseFee,
+		)
+	} else {
+		txs, err := utils.EncodeAndCompressTxList(block.Transactions())
+		if err != nil {
+			return err
+		}
+
+		// convert block to execution envlope
+		log.Info("Publishing L2RequestResponse", "hash", hash.Hex())
+
+		if err := s.p2pNode.GossipOut().PublishL2RequestResponse(ctx, &eth.ExecutionPayloadEnvelope{
+			ExecutionPayload: &eth.ExecutionPayload{
+				BaseFeePerGas: eth.Uint256Quantity(u256),
+				ParentHash:    block.ParentHash(),
+				FeeRecipient:  block.Coinbase(),
+				ExtraData:     block.Extra(),
+				PrevRandao:    eth.Bytes32(block.MixDigest()),
+				BlockNumber:   eth.Uint64Quantity(block.NumberU64()),
+				GasLimit:      eth.Uint64Quantity(block.GasLimit()),
+				GasUsed:       eth.Uint64Quantity(block.GasUsed()),
+				Timestamp:     eth.Uint64Quantity(block.Time()),
+				BlockHash:     block.Hash(),
+				Transactions:  []eth.Data{hexutil.Bytes(txs)},
```

Comment:
@davidtaikocha getting `unexpected EOF` or `invalid header` on zlib decompressing when receiving the block sometimes, if you have time to look at this tonight lmk